gum/arch-x86/gumx86writer.c

```diff
@@ -868,14 +895,43 @@ gum_x86_writer_put_call_reg_offset_ptr (GumX86Writer * self,
   return TRUE;
 }
 
-void
+gboolean
 gum_x86_writer_put_call_indirect (GumX86Writer * self,
-                                  gconstpointer * addr)
+                                  GumAddress address)
 {
-  self->code[0] = 0xff;
-  self->code[1] = 0x15;
-  *((guint32 *) (self->code + 2)) = GUINT32_TO_LE (GUM_ADDRESS (addr));
-  gum_x86_writer_commit (self, 6);
+  if (self->target_cpu == GUM_CPU_AMD64)
+  {
+    gint64 distance = (gssize) address - (gssize) (self->pc + 6);
+
+    if (! GUM_IS_WITHIN_INT32_RANGE (distance))
+      return false;
+
+    self->code[0] = 0xff;
+    self->code[1] = 0x15;
+    *((guint32 *) (self->code + 2)) = GINT32_TO_LE ((gint32) distance);
+    gum_x86_writer_commit (self, 6);
+  }
+  else
+  {
+    self->code[0] = 0xff;
+    self->code[1] = 0x15;
+    *((guint32 *) (self->code + 2)) = GUINT32_TO_LE (address);
+    gum_x86_writer_commit (self, 6);
+  }
+  return true;
+}
+
+gboolean
+gum_x86_writer_put_call_indirect_label (GumX86Writer * self,
+                                        gconstpointer label_id)
+{
+  if (! gum_x86_writer_put_call_indirect(self, self->pc))
+    return false;
+  if (self->target_cpu == GUM_CPU_AMD64)
+    gum_x86_writer_add_label_reference_here (self, label_id, GUM_LREF_NEAR);
+  else
+    gum_x86_writer_add_label_reference_here (self, label_id, GUM_LREF_ABS);
```

Comment:
These lines are long, so a blocks are needed for readability. However, a ternary expression would be clearer here, and would reduce repetition.